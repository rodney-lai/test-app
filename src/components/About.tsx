/**
 *
 * Copyright (c) 2020 Rodney S.K. Lai
 * https://github.com/rodney-lai
 *
 * Permission to use, copy, modify, and/or distribute this software for
 * any purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 */

import * as React from 'react'
import { useState, useEffect } from 'react'
import { useQuery, gql } from '@apollo/client'
import Typography from '@material-ui/core/Typography'
import BuildDate from './BuildDate'
import markdown from 'remark-parse'
import remark2rehype from 'remark-rehype'
import rehype2react from 'rehype-react'
import unified from 'unified'

export const GET_BUILD_DATE = gql`
  query {
    version {
      buildDate
    }
  }
`

interface VersionData {
  buildDate: string
}

interface GraphQLData {
  version: VersionData
}

var processor = unified()
  .use(markdown)
  .use(remark2rehype)
  .use(rehype2react, {createElement: React.createElement})

const About = (props: {show: boolean}) => {
  const { loading, error, data } = useQuery<GraphQLData>(GET_BUILD_DATE)
  const [readmeMarkdown, setReadmeMarkdown] = useState('Loading...')

  useEffect(() => {
    fetch("https://test.rodneylai.com/api/README.md")
      .then(result => result.text())
      .then(result => setReadmeMarkdown(result))
  }, [])

  if (props.show) {
    return(
      <Typography style={{padding:"15px",textAlign:"left"}} variant="subtitle1">
        <>{processor.processSync(readmeMarkdown).result}</>
        {!loading && !error && data &&
          <div>API Build Date: {data.version.buildDate}.</div>
        }
        <div>App Build Date: <BuildDate />.</div>
      </Typography>
    )
  } else {
    return null
  }
}

export default About
