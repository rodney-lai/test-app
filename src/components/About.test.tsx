/**
 *
 * Copyright (c) 2020 Rodney S.K. Lai
 * https://github.com/rodney-lai
 *
 * Permission to use, copy, modify, and/or distribute this software for
 * any purpose with or without fee is hereby granted, provided that the
 * above copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 */

import React from 'react'
import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import About, { GET_BUILD_DATE } from './About'

const mocks = [
  {
    request: {
      query: GET_BUILD_DATE,
    },
    result: {
      data: {
        version: {
          buildDate: "theAPIBuildDate"
        },
      },
    },
  },
]

const server = setupServer(
  rest.get('https://test.rodneylai.com/api/README.md', (req, res, ctx) => {
    return res(ctx.text('This is the readme'))
  })
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

jest.mock('./BuildDate', () => () => "theAppBuildDate")

test('about component when show = true', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <About show={true} />
    </MockedProvider>
  )
  await waitForElementToBeRemoved(() => screen.getByText(/Loading.../i))
  await waitFor(() => screen.getByText(/API Build Date:/i))
  expect(container).toMatchSnapshot()
})

test('about component when show = false', () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <About show={false} />
    </MockedProvider>
  )
  expect(container).toMatchSnapshot()
})
