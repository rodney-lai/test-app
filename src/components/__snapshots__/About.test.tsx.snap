// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`about component when show = false 1`] = `<div />`;

exports[`about component when show = true 1`] = `
<div>
  <h6
    class="MuiTypography-root MuiTypography-subtitle1"
    style="padding: 15px; text-align: left;"
  >
    <div>
      <p>
        This is the readme
      </p>
    </div>
  </h6>
</div>
`;
